lib/Yield/Calls/Slices.cpp

```diff
@@ -0,0 +1,133 @@
+  auto New = std::make_unique<yield::Graph::Node>(Source->data());
+  return Graph.addNode(std::move(New));
+}
+
+/// \tparam NV local `NodeView` specialization
+/// \tparam INV inverted location `NodeView` specialization
+template<typename NV, typename INV>
+yield::Graph
+preservingSliceImpl(const yield::Graph &Input, const MetaAddress &SlicePoint) {
+  auto LookupHelper = [&SlicePoint](NodeView Node) {
+    return Node->Address == SlicePoint;
+  };
+  auto Entry = llvm::find_if(Input.nodes(), LookupHelper);
+  revng_assert(Entry != Input.nodes().end());
+
+  // Find the rank of each node, such that for any node its rank is equal to
+  // the highest rank among its children plus one.
+  std::unordered_map<const yield::Graph::Node *, size_t> Ranks;
+  std::vector<NodeView> PostOrder(llvm::po_begin(NV{ *Entry }),
+                                  llvm::po_end(NV{ *Entry }));
+  for (NodeView CurrentNode : llvm::reverse(PostOrder)) {
+    uint64_t &CurrentRank = Ranks[CurrentNode];
+
+    CurrentRank = 0;
+    for (auto Child : llvm::children<INV>(CurrentNode))
+      if (auto RankIterator = Ranks.find(Child); RankIterator != Ranks.end())
+        CurrentRank = std::max(RankIterator->second + 1, CurrentRank);
+  }
+
+  // For each node, select a single predecessor to keep connected to.
+  // The ranks calculated earlier are used to choose a specific one.
+  //
+  // TODO: We should consider a better selection algorithm.
+  std::unordered_map<NodeView, NodeView> RealEdges;
+  for (NodeView Node : Input.nodes()) {
+    NodeView SelectedNeighbour = nullptr;
+
+    // This selects a child with the highest rank
+    auto NodeIt = Ranks.find(Node);
+    if (NodeIt != Ranks.end()) {
+      size_t SelectedNeighbourRank = 0;
+      for (NodeView Neighbour : llvm::children<INV>(Node)) {
+        auto NeighbourIt = Ranks.find(Neighbour);
+        if (NeighbourIt != Ranks.end()) {
+
+          // Select the neighbour with the highest possible rank that is still
+          // lower than the current node's rank.
+          size_t Rank = NeighbourIt->second;
+          if (Rank < NodeIt->second && Rank >= SelectedNeighbourRank) {
+            SelectedNeighbour = Neighbour;
+            SelectedNeighbourRank = Rank;
+          }
+        }
+      }
+    }
+
+    auto [_, Success] = RealEdges.try_emplace(Node, SelectedNeighbour);
+    revng_assert(Success);
+  }
+
+  yield::Graph Result;
+  std::unordered_map<NodeView, yield::Graph::Node *> Lookup;
+
+  // Returns the version of the node from the new graph if it exists,
+  // or adds a new one to if it does not.
+  auto FindOrAddHelper = [&Result, &Lookup](NodeView OldNode) {
+    if (auto NewNode = Lookup.find(OldNode); NewNode != Lookup.end())
+      return NewNode->second;
+    else
+      return Lookup.emplace(OldNode, nodeHelper(Result, OldNode)).first->second;
+  };
+
+  // Make sure entry node is always present so that this never returns an empty
+  // graph
+  FindOrAddHelper(*Entry);
```

Comment:
This looks redundant to me. This entire function should assert that `SlicePoint` is found. At that point you never need to call `nodeHelper` to emit the entry anymore, because you have the guarantee that the entry point is present and it is the node associated with `SlicePoint`.

Reply:
This calls guarantees that the `SlicePoint` is present in the output graph. Consider the case where the input graph is a single node - no loops after this one ever do anything (there are no edges).

Reply:
Should I expand the comment to explain this better?

Reply:
What do you think? https://github.com/revng/revng/pull/191/commits/bec0c764f40b1864f9d62467d45cc22357e3bedd